libcloud/common/digitalocean.py

```diff
@@ -0,0 +1,227 @@
+    'DigitalOceanBaseDriver'
+]
+
+
+class DigitalOcean_v1_Response(JsonResponse):
+    def parse_error(self):
+        if self.status == httplib.FOUND and '/api/error' in self.body:
+            # Hacky, but DigitalOcean error responses are awful
+            raise InvalidCredsError(self.body)
+        elif self.status == httplib.UNAUTHORIZED:
+            body = self.parse_body()
+            raise InvalidCredsError(body['message'])
+        else:
+            body = self.parse_body()
+
+            if 'error_message' in body:
+                error = '%s (code: %s)' % (body['error_message'], self.status)
+            else:
+                error = body
+            return error
+
+
+class DigitalOcean_v1_Connection(ConnectionUserAndKey):
+    """
+    Connection class for the DigitalOcean (v1) driver.
+    """
+
+    host = 'api.digitalocean.com'
+    responseCls = DigitalOcean_v1_Response
+
+    def add_default_params(self, params):
+        """
+        Add parameters that are necessary for every request
+
+        This method adds ``client_id`` and ``api_key`` to
+        the request.
+        """
+        params['client_id'] = self.user_id
+        params['api_key'] = self.key
+        return params
+
+
+class DigitalOcean_v2_Response(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                            httplib.NO_CONTENT]
+
+    def parse_error(self):
+        if self.status == httplib.UNAUTHORIZED:
+            body = self.parse_body()
+            raise InvalidCredsError(body['message'])
+        else:
+            body = self.parse_body()
+            if 'message' in body:
+                error = '%s (code: %s)' % (body['message'], self.status)
+            else:
+                error = body
+            return error
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class DigitalOcean_v2_Connection(ConnectionKey):
+    """
+    Connection class for the DigitalOcean (v2) driver.
+    """
+
+    host = 'api.digitalocean.com'
+    responseCls = DigitalOcean_v2_Response
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+
+        This method adds ``token`` to the request.
+        """
+        headers['Authorization'] = 'Bearer %s' % (self.key)
+        headers['Content-Type'] = 'application/json'
+        return headers
+
+
+class DigitalOceanConnection(DigitalOcean_v2_Connection):
+    """
+    Connection class for the DigitalOcean driver.
+    """
+    pass
+
+
+class DigitalOceanResponse(DigitalOcean_v2_Response):
+    pass
+
+
+class DigitalOceanBaseDriver(BaseDriver):
+    """
+    DigitalOcean BaseDriver
+    """
+    name = 'DigitalOcean'
+    website = 'https://www.digitalocean.com'
+
+    def __new__(cls, key, secret=None, api_version='v2', **kwargs):
+        if cls is DigitalOceanBaseDriver:
+            if api_version == 'v1' or secret is not None:
```

Comment:
To avoid surprises we should throw if user specifies api_version == 'v2' and secret or api_version == 'v1' and missing secret).